pilot/pkg/networking/core/v1alpha3/fake.go

```diff
@@ -0,0 +1,272 @@
+	se := serviceentry.NewServiceDiscovery(configController, model.MakeIstioStore(configStore), &FakeXdsUpdater{})
+	// TODO allow passing in registry, for k8s, mem reigstry
+	serviceDiscovery.AddRegistry(se)
+	msd := memregistry.NewServiceDiscovery(opts.Services)
+	msd.ClusterID = string(serviceregistry.Mock)
+	serviceDiscovery.AddRegistry(serviceregistry.Simple{
+		ClusterID:        string(serviceregistry.Mock),
+		ProviderID:       serviceregistry.Mock,
+		ServiceDiscovery: msd,
+		Controller:       msd.Controller,
+	})
+	for _, reg := range opts.ServiceRegistries {
+		serviceDiscovery.AddRegistry(reg)
+	}
+
+	env := &model.Environment{}
+	env.PushContext = model.NewPushContext()
+	env.ServiceDiscovery = serviceDiscovery
+	env.IstioConfigStore = model.MakeIstioStore(configStore)
+	env.Watcher = mesh.NewFixedWatcher(m)
+	if opts.NetworksWatcher == nil {
+		opts.NetworksWatcher = mesh.NewFixedNetworksWatcher(nil)
+	}
+	env.NetworksWatcher = opts.NetworksWatcher
+
+	// Setup configuration. This should be done after registries are added so they can process events.
+	for _, cfg := range configs {
+		if _, err := configStore.Create(cfg); err != nil {
+			t.Fatalf("failed to create config %v: %v", cfg.Name, err)
+		}
+	}
+
+	// TODO allow passing event handlers for controller
+
+	retry.UntilSuccessOrFail(t, func() error {
+		if !serviceDiscovery.HasSynced() {
+			return errors.New("not synced")
+		}
+		return nil
+	})
+
+	se.ResyncEDS()
+	if err := env.PushContext.InitContext(env, nil, nil); err != nil {
+		t.Fatalf("Failed to initialize push context: %v", err)
+	}
+
+	if opts.Plugins == nil {
+		opts.Plugins = registry.NewPlugins([]string{plugin.Authn, plugin.Authz})
+	}
+
+	fake := &ConfigGenTest{
+		t:                    t,
+		store:                configController,
+		env:                  env,
+		ConfigGen:            NewConfigGenerator(opts.Plugins),
+		MemRegistry:          msd,
+		ServiceEntryRegistry: se,
+		pushContextLock:      opts.PushContextLock,
+	}
+	return fake
+}
+
+// SetupProxy initializes a proxy for the current environment. This should generally be used when creating
+// any proxy. For example, `p := SetupProxy(&model.Proxy{...})`.
+func (f *ConfigGenTest) SetupProxy(p *model.Proxy) *model.Proxy {
+	// Setup defaults
+	if p == nil {
+		p = &model.Proxy{}
+	}
+	if p.Metadata == nil {
+		p.Metadata = &model.NodeMetadata{}
+	}
+	if p.Metadata.IstioVersion == "" {
+		p.Metadata.IstioVersion = "1.8.0"
+		p.IstioVersion = model.ParseIstioVersion(p.Metadata.IstioVersion)
+	}
+	if p.Type == "" {
+		p.Type = model.SidecarProxy
+	}
+	if p.ID == "" {
+		p.ID = "app.test"
+	}
+	if len(p.IPAddresses) == 0 {
+		p.IPAddresses = []string{"1.1.1.1"}
+	}
+
+	// Initialize data structures
+	pc := f.PushContext()
+	p.SetSidecarScope(pc)
+	p.SetGatewaysForProxy(pc)
+	if err := p.SetServiceInstances(f.env.ServiceDiscovery); err != nil {
+		f.t.Fatal(err)
+	}
+	p.DiscoverIPVersions()
+	return p
+}
+
+// TODO do we need lock around push context?
```

Comment:
For only a few scenarios, like PushContext re-init when meshNetworks change, for example.

Reply:
Oops this is a stale TODO (and there may be others). I added the lock for this reason. thanks, will clean up